src/MongoDB/Logging/functions.c

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2016-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "bson/bson.h"
+#include "mongoc/mongoc.h"
+
+#include <php.h>
+
+#include "php_phongo.h"
+#include "phongo_error.h"
+#include "phongo_log.h"
+
+#define PHONGO_LOG_DOMAIN "php"
+
+/* Registers a global logger */
+PHP_FUNCTION(MongoDB_Driver_Logging_addLogger)
+{
+	zval* logger;
+
+	PHONGO_PARSE_PARAMETERS_START(1, 1)
+	Z_PARAM_OBJECT_OF_CLASS(logger, php_phongo_logger_ce)
+	PHONGO_PARSE_PARAMETERS_END();
+
+	phongo_log_add_logger(logger);
+}
+
+/* Log a message */
+PHP_FUNCTION(MongoDB_Driver_Logging_log)
+{
+	zend_long level;
+	char*     message;
+	size_t    message_len;
+
+	PHONGO_PARSE_PARAMETERS_START(2, 2)
+	Z_PARAM_LONG(level)
+	Z_PARAM_STRING(message, message_len)
+	PHONGO_PARSE_PARAMETERS_END();
+
+	/* TODO: throw if level is invalid */
+
+	if (strlen(message) != message_len) {
+		phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE, "Log messages cannot contain null bytes. Unexpected null byte after \"%s\".", message);
+		return;
+	}
+
+	mongoc_log(level, PHONGO_LOG_DOMAIN, "%s", message);
```

Comment:
Not yet a C expert like you, but I think the `mongoc_log` function accepts extra arguments. That looks like the [PSR-3 log functions](https://www.php-fig.org/psr/psr-3/#3-psrlogloggerinterface) that accept a `$context` that is injected into the message.

https://github.com/mongodb/mongo-c-driver/blob/master/src/libmongoc/src/mongoc/mongoc-log.c#L115C1-L115C1

I would be nice if `MongoDB\Driver\Logging\Logger::log` could receive this raw `format` and `args`. Log systems that accept objects are becoming increasingly common (I'm thinking of kibana, which lets you filter on log fields).

Also, would it be possible to pass the manager (or a Client identifier) that triggered the log message?

Reply:
> I think the mongoc_log function accepts extra arguments.

It does, but the variable args are only used for the format string (see: [Logging](
https://mongoc.org/libmongoc/current/logging.html) in the libmongoc docs). I'd rather not expose that in in PHPC when userland can filter messages through [`sprintf()`](https://www.php.net/sprintf) and get much better error reporting.

If you were referring to `gLogData` in the C source, that's an arbitrary pointer provided when registering a handler, which libmongoc then passes to the handler function with each invocation. That wouldn't be suitable for 

> Log systems that accept objects are becoming increasingly common

I think that's what [structured logging](https://github.com/mongodb/specifications/blob/master/source/logging/logging.rst#structured-versus-unstructured-logging) in the driver spec refers to, but it's not supported in libmongoc. And I don't think we need it ourselves, since PHPC and PHPLIB will likely only use logging for messages mandated by our own specifications.

> Also, would it be possible to pass the manager (or a Client identifier) that triggered the log message?

Logging in libmongoc is fundamentally disconnected from `mongoc_client_t` and only exists at a global level. Some drivers include a topology ID (from SDAM) in logged messages to address that; however, libmongoc does not (yet) do so.

Reply:
@GromNaN: OK to resolve this?

Reply:
Yes, thanks.